### PR TITLE
uniter: Be smarter about when we run config-changed (part 1)

### PR DESCRIPTION
**Pull request by babbageclunk**
## Description of change
At the moment, when a controller agent restarts all of the unit agents that were connected to it reconnect and then immediately fire config-changed, even if no configuration has actually changed. On controllers with thousands of units, this can overload the controller agent while all the units get all their config and run any other hook commands in response.

The uniter always runs config-changed when it first starts because it doesn't track any local state that would enable it to see that the config is the same and avoid running the hook.

Change the API to expose Uniter.WatchConfigSettingsHash - this works like .WatchConfigSettings, except it sends a hash of the current config when it signals. The uniter stores that value in the local uniter state file, so that it persists across agent restarts, and checks it before firing config-changed.

This bumps the uniter API to version 9.

This PR plumbs the config hash through the API and uniter, but the uniter also watches trust config and addresses, so the behaviour won't change until those other parts are done in followups.

## QA steps

* Bootstrap a controller and model, deploy some applications and relate them. 
* See that the uniter state files for various units includes the config-hash.
* Change a config value for one of the applications.
* See that the config-hash changes when the config-changed hook fires.

## Documentation changes

None

### CODE REVIEW

**Comment by babbageclunk** [comment]
Note to self - work out whether/how the uniter restarts the WatchConfigSettings(Hash) watcher when the application's charm is updated. If it doesn't it'll need to!

File: worker/uniter/remotestate/watcher_test.go [hunk 1]
```diff
@@ -188,9 +188,9 @@ func (s *WatcherSuiteIAAS) TestSnapshot(c *gc.C) {
 	s.signalAll()
 	assertNotifyEvent(c, s.watcher.RemoteStateChanged(), "waiting for remote state change")
 
-	// Note that the configuration change is updated on both application and
-	// charm config which increments it twice.
-	expectedVersion := 3
+	// Note that the configuration change is updated on both application config and
```
Comment (wallyworld) [inline worker/uniter/remotestate/watcher_test.go]: I think this should be " on both trust changes and address changes "
The uniter watches for trust changes, and on the server side it's done in terms of app config changes.

File: worker/uniter/remotestate/watcher_test.go [hunk 2]
```diff
@@ -211,9 +212,9 @@ func (s *WatcherSuiteCAAS) TestSnapshot(c *gc.C) {
...
```
Comment (wallyworld) [inline worker/uniter/remotestate/watcher_test.go]: ditto

File: worker/uniter/resolver_test.go [hunk 3]
```diff
@@ -368,3 +368,57 @@ func (s *resolverSuite) TestRunHookStopRetryTimer(c *gc.C) {
 	c.Assert(err, gc.Equals, resolver.ErrNoOperation)
 	s.stub.CheckCallNames(c, "StartRetryHookTimer", "StopRetryHookTimer")
 }
+
+func (s *resolverSuite) TestRunsConfigChangedIfConfigHashChanges(c *gc.C) {
```
Comment (wallyworld) [inline worker/uniter/resolver_test.go]: A test for no op if hashes haven't changed?

File: apiserver/facades/agent/uniter/uniter.go [hunk 4]
```diff
@@ -2673,3 +2693,59 @@ func (u *UniterAPI) goalStateUnits(app *state.Application, principalName string)
...
+	return result, nil
+}
+
+// Mask WatchConfigSettingsHash from the v8 API. The API reflection
+// code in rpc/rpcreflect/type.go:newMethod skips 2-argument methods,
+// so this removes the method as far as the RPC machinery is
+// concerned.
+
+// WatchConfigSettingsHash isn't on the v8 API.
+func (u *UniterAPIV8) WatchConfigSettingsHash(_, _ struct{}) {}
```
[the remark this answers is not in the document]
Comment (babbageclunk) [reply apiserver/facades/agent/uniter/uniter.go]: I'd like to do that in a separate PR - once I go through and make sure nothing else uses it.

File: state/unit_test.go [hunk 5]
```diff
@@ -165,6 +165,118 @@ func (s *UnitSuite) TestWatchConfigSettings(c *gc.C) {
...
+	// Add an attribute that comes before.
+	err = s.application.UpdateCharmConfig(charm.Settings{
+		"alphabetic": 1,
+	})
+	c.Assert(err, jc.ErrorIsNil)
+	// Sanity check.
+	config, err := s.unit.ConfigSettings()
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(config, gc.DeepEquals, charm.Settings{
+		// Ints that get round-tripped through mongo come back as int64.
```
[the remark this answers is not in the document]
Comment (babbageclunk) [reply state/unit_test.go]: Good call, I'll add a test for that.

File: worker/uniter/remotestate/watcher_test.go (hunk 1 above)
Comment (babbageclunk) [reply worker/uniter/remotestate/watcher_test.go]: Right, that makes sense.

File: worker/uniter/resolver_test.go (hunk 3 above)
Comment (babbageclunk) [reply worker/uniter/resolver_test.go]: Ok, I'll add that.

**Comment by babbageclunk** [comment]
I can confirm that the hash keeps getting updated if I change a config value after a charm upgrade, so the uniter must be recreating the config settings hash watcher. I'll make sure I check the same thing happens when the other two hashes are done.

**Comment by babbageclunk** [comment]
$$merge$$

**Pull request merged**